### PR TITLE
[clang-tidy] Add support for determining constness of more expressions.

### PR DESCRIPTION
**Pull request by legrosbuffle**
This uses a more systematic approach for determining whcich `DeclRefExpr`s mutate the underlying object: Instead of using a few matchers, we walk up the AST until we find a parent that we can prove cannot change the underlying object.

This allows us to handle most address taking and dereference, bindings to value and const& variables, and track constness of pointee (see changes in DeclRefExprUtilsTest.cpp).

This allows supporting more patterns in `performance-unnecessary-copy-initialization`.

Those two patterns are relatively common:

```
const auto e = (*vector_ptr)[i]
```

and

```
const auto e = vector_ptr->at(i);
```

In our codebase, we have around 25% additional findings from `performance-unnecessary-copy-initialization` with this change. I did not see any additional false positives.

### CODE REVIEW

File: clang-tools-extra/clang-tidy/utils/DeclRefExprUtils.cpp [hunk 1]
```diff
@@ -34,69 +36,185 @@ void extractNodesByIdTo(ArrayRef<BoundNodes> Matches, StringRef ID,
...
+    const auto Parents = Ctx.getParents(*Entry.E);
+    // Note: most nodes have a single parents, but there exist nodes that have
+    // several parents, such as `InitListExpr` that have semantic and syntactic
+    // forms.
+    for (const auto &Parent : Parents) {
+      if (Parent.get<CompoundStmt>()) {
+        // Unused block-scope statement.
+        continue;
+      }
+      const Expr *const P = Parent.get<Expr>();
```
Comment (EugeneZelenko) [inline clang-tools-extra/clang-tidy/utils/DeclRefExprUtils.cpp:108]: `auto` could be used here - type is explicitly stated.

File: clang-tools-extra/clang-tidy/utils/DeclRefExprUtils.cpp [hunk 2]
```diff
@@ -34,69 +36,185 @@ void extractNodesByIdTo(ArrayRef<BoundNodes> Matches, StringRef ID,
...
+      assert(Ty->isPointerType());
+      Ty = Ty->getPointeeType().getCanonicalType();
+    }
+    if (Ty.isConstQualified()) {
+      continue;
+    }
+
+    // Otherwise we have to look at the parents to see how the expression is
+    // used.
+    const auto Parents = Ctx.getParents(*Entry.E);
```
Comment (EugeneZelenko) [inline clang-tools-extra/clang-tidy/utils/DeclRefExprUtils.cpp:99]: Please don't use `auto` - type is not spelled.

**Review by EugeneZelenko — COMMENTED** [review]
Please mention changes in Release Notes.

**Review by fberger — APPROVED** [review]

**Review by PiotrZSL — APPROVED** [review]
I didn't check entire change, in some places `{}` in if's aren't needed.
Overall looking fine, but this check need to be run on bigger code-base to verify if there are no false positives or crashes.

**Comment by legrosbuffle** [comment]
Thanks all. Comments addressed.

> Overall looking fine, but this check need to be run on bigger code-base to verify if there are no false positives or crashes.

Sorry, this was not very clear. By "our codebase" I meant the whole of Google code :) This also ran on numerous third party repos.

**Pull request merged**